src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs

```diff
@@ -32,29 +32,58 @@ public class TestJsonCommand : PSCmdlet
         /// then validates the JSON against the schema. Before testing the JSON string,
         /// the cmdlet parses the schema doing implicitly check the schema too.
         /// </summary>
-        [Parameter(Position = 1)]
+        [Parameter(Position = 1, ParameterSetName = TestJsonCommand.SchemaStringParameterSet)]
         [ValidateNotNullOrEmpty()]
         public string Schema { get; set; }
 
+        /// <summary>
+        /// A path to the file containg schema to validate the JSON against.
+        /// This is optional parameter.
+        /// If the parameter is absent the cmdlet only attempts to parse the JSON string.
+        /// If the parameter present the cmdlet attempts to parse the JSON string and
+        /// then validates the JSON against the schema. Before testing the JSON string,
+        /// the cmdlet parses the schema doing implicitly check the schema too.
+        /// </summary>
+        [Parameter(Position = 1, ParameterSetName = TestJsonCommand.SchemaPathParameterSet)]
+        [ValidateNotNullOrEmpty()]
+        public string SchemaPath { get; set; }
+
         private JsonSchema _jschema;
 
         /// <summary>
         /// Prepare an JSON schema.
         /// </summary>
         protected override void BeginProcessing()
         {
-            if (Schema != null)
+            try
             {
-                try
+                if (Schema != null)
                 {
                     _jschema = JsonSchema.FromJsonAsync(Schema).Result;
+
                 }
-                catch (Exception exc)
+                else if (SchemaPath != null)
                 {
-                    Exception exception = new Exception(TestJsonCmdletStrings.InvalidJsonSchema, exc);
-                    ThrowTerminatingError(new ErrorRecord(exception, "InvalidJsonSchema", ErrorCategory.InvalidData, null));
+                    string resolvedpath = string.Empty;
+
+                    try
+                    {
+                        resolvedpath = PathUtils.ResolveFilePath(SchemaPath, this, true);
+                    }
+                    catch (ItemNotFoundException e)
+                    {
+                        // NOTE: This will throw
+                        PathUtils.ReportFileOpenFailure(this, resolvedpath, e);
```

Comment:
Are you sure that PathUtils.ResolveFilePath() does not do this?

Reply:
I wasn't sure what is the idiomatic way of resolving file path, so I used the code from the [Send-MailMessage.cs#L320-L335](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs#L320-L335).

Reply:
Seems that `ResolveFilePath` doesn't throw in any conditions I could test.
`NJsonSchema` will throw `System.IO.IOException` so I suppose we catch that.

P.S. Does creating exception handler for `System.IO.IOException` and throwing more specific error makes sense, or it's fine to assume that user will inspect `$Error` to get inner exception? This will probably require new resources in the `TestJsonCmdletStrings.resx`.

Reply:
`ResolveFilePath` parses a string as a possible path but doesn't check file existence or access. I think we could use pattern from Get-FileHash cmdlet (LiteralPath).

Reply:
Thanks! I've had a look at `Get-FileHash` but while the pattern looks good, handling exceptions from async methods is PITA. I tried to do my best, but I'm not sure if its a correct approach. Could you give it another look? I've left a handful of comments in the code, I hope that helps.

Reply:
How many exceptions could be in the aggregate exception? If many you could look Add-Type cmdlet. There we get errors from compiler and write them to output, then we write generic terminating error.

Reply:
In practice it's one exception, but we cannot guarantee that. The main problem I've tried to solve here is that in the case of `JsonSchema.FromFileAsync` it would be nested like this: `AggregateException` → `TargetInvocationException` → `IOException`. If we just let it bubble up, user would have to do `$Error[0].Exception.InnerException.InnerException.InnerException` just to get to the error details in this case. If we don't care about this, than we can just handle exception the way we did before.

Reply:
I am ok with unwrapping the AggregateException.

Reply:
>  It seems it could be used twice so you could implement a private method.

Yeah, that was my intention if unwrapping is ok. Thanks.

Reply:
I've converted delegates to a private method.